### PR TITLE
8254719: ZGC: Clean up includes

### PR DESCRIPTION
**Pull request by pliden**
I noticed quite a few includes in ZGC that are no longer needed and can be cleaned out. As a side-effect I noticed that `stackWatermark.hpp` and `threadSMR.hpp` weren't self contained, so I had to fix that too.

Builds on all Oracle platforms.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8254719](https://bugs.openjdk.java.net/browse/JDK-8254719): ZGC: Clean up includes

### Reviewers
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/644/head:pull/644`
`$ git checkout pull/644`

### CODE REVIEW

File: src/hotspot/share/gc/z/zHeap.hpp [hunk 1]
```diff
@@ -39,6 +38,7 @@
 #include "gc/z/zUnload.hpp"
 #include "gc/z/zWorkers.hpp"
 
+class Page;
```
Comment (stefank) [inline src/hotspot/share/gc/z/zHeap.hpp:41]: Did you mean ZPage?

File: src/hotspot/share/gc/z/zMark.hpp [hunk 2]
```diff
@@ -26,19 +26,17 @@
 
 #include "gc/z/zMarkStack.hpp"
 #include "gc/z/zMarkStackAllocator.hpp"
+#include "gc/z/zMarkStackEntry.hpp"
 #include "gc/z/zMarkTerminate.hpp"
 #include "oops/oopsHierarchy.hpp"
-#include "utilities/globalDefinitions.hpp"
```
Comment (stefank) [inline src/hotspot/share/gc/z/zMark.hpp:31]: This file uses definitions brought by globalDefinitions.hpp, so I don't see the reason for removing it.

File: src/hotspot/share/gc/z/zRelocate.cpp [hunk 3]
```diff
@@ -27,14 +27,12 @@
 #include "gc/z/zForwarding.inline.hpp"
 #include "gc/z/zHeap.hpp"
 #include "gc/z/zOopClosures.inline.hpp"
-#include "gc/z/zPage.hpp"
```
Comment (stefank) [inline src/hotspot/share/gc/z/zRelocate.cpp:30]: This file calls `forwarding->page()->object_iterate(&cl);` so it *should* be including zPage.inline.hpp

File: src/hotspot/share/gc/z/zRelocate.hpp [hunk 4]
```diff
@@ -24,11 +24,11 @@
 #ifndef SHARE_GC_Z_ZRELOCATE_HPP
 #define SHARE_GC_Z_ZRELOCATE_HPP
 
-#include "gc/z/zRelocationSet.hpp"
 #include "gc/z/zWorkers.hpp"
```
Comment (stefank) [inline src/hotspot/share/gc/z/zRelocate.hpp:27]: This could also be forward declared, right?

File: src/hotspot/share/gc/z/zRelocationSet.cpp [hunk 5]
```diff
@@ -24,7 +24,6 @@
 #include "precompiled.hpp"
 #include "gc/z/zForwarding.hpp"
 #include "gc/z/zRelocationSet.hpp"
-#include "memory/allocation.hpp"
```
Comment (stefank) [inline src/hotspot/share/gc/z/zRelocationSet.cpp:27]: Uses REALLOC_C_HEAP_ARRAY, so probably should keep allocation.hpp

File: src/hotspot/share/gc/z/zRelocationSetSelector.cpp [hunk 6]
```diff
@@ -30,7 +30,6 @@
 #include "logging/log.hpp"
 #include "runtime/globals.hpp"
 #include "utilities/debug.hpp"
-#include "utilities/powerOfTwo.hpp"
```
Comment (stefank) [inline src/hotspot/share/gc/z/zRelocationSetSelector.cpp:33]: Uses exact_log2, so shouldn't be removed.

**Review by stefank — CHANGES_REQUESTED** [review]
I think there are too many removal of includes that shouldn't be removed, IMHO. Could you take another pass and check all the files? I commented on some of the files, but didn't look at them all.

**Comment by pliden** [comment]
@stefank Thanks for reviewing. I went through all files again, addressed your comments, and found a few additional missing includes.

File: src/hotspot/share/gc/z/zVerify.cpp [hunk 7]
```diff
@@ -33,9 +33,9 @@
 #include "gc/z/zStat.hpp"
 #include "gc/z/zVerify.hpp"
 #include "memory/iterator.inline.hpp"
-#include "memory/resourceArea.inline.hpp"
-#include "oops/oop.hpp"
```
Comment (stefank) [inline src/hotspot/share/gc/z/zVerify.cpp:37]: This file uses oopDesc::is_oop, so I think oops/oop.hpp should stay.

**Review by stefank — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/z/zVerify.cpp (hunk 7 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zVerify.cpp:37]: True. Also noticed this file was, from the beginning, missing a few other includes, so I fixed those too.

**Review by stefank — APPROVED** [review]

**Pull request closed without merging**